securedrop/journalist_app/api2/types.py

```diff
@@ -74,63 +79,132 @@ class Target:
 
     version: Version
 
+    def __post_init__(self) -> None:
+        v = str(self.version)
 
-@dataclass
+        if len(v) != VERSION_LEN:
+            raise ValueError(f"version must have {VERSION_LEN} hex digits")
+
+        try:
+            int(v, 16)
+        except ValueError:
+            raise ValueError("version must be hex-encoded")
+
+
+@dataclass(frozen=True)
 class SourceTarget(Target):
     source_uuid: SourceUUID
 
+    def __post_init__(self) -> None:
+        super().__post_init__()
+        if not self.source_uuid:
+            raise ValueError("source_uuid must be non-empty")
 
-@dataclass
+        try:
+            UUID(str(self.source_uuid))
+        except ValueError:
+            raise ValueError(f"invalid source UUID: {self.source_uuid}")
+
+
+@dataclass(frozen=True)
 class ItemTarget(Target):
     item_uuid: ItemUUID
 
+    def __post_init__(self) -> None:
+        super().__post_init__()
+        if not self.item_uuid:
+            raise ValueError("item_uuid must be non-empty")
 
-@dataclass
+        try:
+            UUID(str(self.item_uuid))
+        except ValueError:
+            raise ValueError(f"invalid item UUID: {self.item_uuid}")
+
+
+@dataclass(frozen=True)
 class EventData:
     """
     Base class for `<EventType>Data dataclasses, to make their union usable at runtime.
     For non-empty events, subclass and add to `EVENT_DATA_TYPES`.
     """
 
 
-@dataclass
+@dataclass(frozen=True)
 class ReplySentData(EventData):
     uuid: ReplyUUID
     reply: str
 
+    def __post_init__(self) -> None:
+        try:
+            UUID(str(self.uuid))
+        except ValueError:
+            raise ValueError(f"invalid reply UUID: {self.uuid}")
+
+        if not self.reply:
+            raise ValueError("reply must be a non-empty string")
+
 
 EVENT_DATA_TYPES = {EventType.REPLY_SENT: ReplySentData}
 
 
-@dataclass
+@dataclass(frozen=True)
 class Event:
     id: EventID
-    target: Target | Mapping
+    target: Target | Mapping[str, Any]
     type: EventType
-    data: Optional[EventData | Mapping] = None
+    data: Optional[EventData | Mapping[str, Any]] = None
 
     def __post_init__(self) -> None:
-        if not isinstance(self.type, EventType):
-            self.type = EventType(self.type)  # strict enum
+        # ID must be usable as an int (for snowflake ordering):
+        if not str(self.id).isdigit():
```

Comment:
I thought we made the event_id typed as a string to make it more opaque (is that wrong?)? If so, then why are we validating that it's all digits here?

Reply:
Yes, both!

https://github.com/freedomofpress/securedrop/blob/42cdb9ef624e391f8df61df4a33caeac3bc1e639/API2.md?plain=1#L222-L233

Clarified in 6ce810bafcef7d5f01dd1d99297057bef8b0bf57.